pkg/render/manager_route_config.go

```diff
@@ -0,0 +1,396 @@
+	// Path is the path portion of the URL based on which we proxy
+	Path string `json:"path"`
+
+	// CABundlePath is where we read the CA bundle from to authenticate the
+	// destination (if non-empty)
+	CABundlePath string `json:"caBundlePath,omitempty"`
+	// PathRegexp, if not nil, checks if Regexp matches the path
+	PathRegexp string `json:"pathRegexp,omitempty"`
+	// PathReplace if not nil will be used to replace PathRegexp matches
+	PathReplace string `json:"pathReplace,omitempty"`
+	// AllowInsecureTLS allows https with insecure tls settings
+	AllowInsecureTLS bool `json:"allowInsecureTLS,omitempty"`
+
+	// ClientCertPath and ClientKeyPath can be set for mTLS on the connection
+	// from Voltron to the destination.
+	ClientCertPath string `json:"clientCertPath,omitempty"`
+	ClientKeyPath  string `json:"clientKeyPath,omitempty"`
+
+	Unauthenticated bool `json:"unauthenticated,omitempty"`
+}
+
+type tlsPassThroughRoute struct {
+	// Destination is the destination URL
+	Destination string `json:"destination"`
+
+	// ServerName
+	ServerName string `json:"serverName"`
+}
+
+// VoltronRouteConfigBuilder is an interface that provides methods to build a VoltronRouteConfig.
+// Implementations of this interface should provide methods to add different types of routes,
+// such as TLSTerminatedRoute and TLSPassThroughRoute, as well as methods to add ConfigMaps
+// and Secrets to the route configuration to generate annotations that change when the contents of
+// objects change (used to restart Voltron). The Build method should be used to create the final
+// VoltronRouteConfig object.
+type VoltronRouteConfigBuilder interface {
+	AddTLSTerminatedRoute(routes operatorv1.TLSTerminatedRoute)
+	AddTLSPassThroughRoute(routes operatorv1.TLSPassThroughRoute)
+	AddConfigMap(configMap *corev1.ConfigMap)
+	AddSecret(secret *corev1.Secret)
+	Build() (*VoltronRouteConfig, error)
+}
+
+type voltronRouteConfigBuilder struct {
+	mountedConfigMaps map[string]struct{}
+	mountedSecrets    map[string]struct{}
+
+	volumeMounts    []corev1.VolumeMount
+	volumes         []corev1.Volume
+	hashableContent []interface{}
+
+	configMaps map[string]*corev1.ConfigMap
+	secrets    map[string]*corev1.Secret
+
+	tlsTerminatedRoutes  []operatorv1.TLSTerminatedRoute
+	tlsPassThroughRoutes []operatorv1.TLSPassThroughRoute
+}
+
+func NewVoltronRouteConfigBuilder() VoltronRouteConfigBuilder {
+	return &voltronRouteConfigBuilder{
+		configMaps:        map[string]*corev1.ConfigMap{},
+		secrets:           map[string]*corev1.Secret{},
+		mountedConfigMaps: map[string]struct{}{},
+		mountedSecrets:    map[string]struct{}{},
+	}
+}
+
+func (builder *voltronRouteConfigBuilder) AddConfigMap(configMap *corev1.ConfigMap) {
+	if _, ok := builder.configMaps[configMap.Name]; !ok {
+		builder.configMaps[configMap.Name] = configMap
+	}
+}
+
+func (builder *voltronRouteConfigBuilder) AddSecret(secret *corev1.Secret) {
+	if _, ok := builder.secrets[secret.Name]; !ok {
```

Comment:
What does it mean if this is called twice on the same secret? Should we error if that's the case? Probably suggests a bug somewhere else in the code, yeah?

Reply:
The same secret can be referenced by multiple tls routes, or even with the same tls route (i.e. if the mtls cert and key are both in the same secret).

So instead of adding check if exists logic at a higher level and complicating the controllers use of the builder (as it would need to track this information) we just ignore duplicate secrets.